Wodes grimoire: Initial Release

BloodSpartan opened the pull request:
Adds 10 new spells + their icons

I think I did this right and I'm glad to finally be able to publish this!

BloodSpartan committed:
Wodes_Grimoire: Initial Release

Adds 10 spells + icons

BloodSpartan committed:
Wodes_Grimoire: remove unneeded file

BloodSpartan committed:
Wodes_Grimoire: Update Screenshot

jordan-octavia-blue committed:
Wode's Grimoire: fix Grace

Simplified Grace's effect function.
The addModifier gets called for both prediction and regular so it
doesn't need to be wrapped in an if function checking for prediction.
The sfx however does, and the sfx should only be called once since
it should play only once no matter how many targets there are.

modifierGraceCountdown and castingGracecounter (renamed for clarity to
graceCountdown and graceQuantity respectively) need to be a part of
the modifiers state, otherwise they are global to the whole module.

BloodSpartan committed:
Wodes_Grimoire: Cleaned up modifier spells

BloodSpartan committed:
Merge branch 'Wodes_Grimoire' of https://github.com/BloodSpartan/spellmasons-mods into Wodes_Grimoire

BloodSpartan commented:
Most modifier spells have addModifier on both client and server, so they could be merged. Regenerate needed to have its variables set into the modifier instead of being in the global module.

BloodSpartan committed:
Wode's Grimoire: Adds mana trails to Harvest

Cleans up loose comments and adds floating text aswell.

BloodSpartan committed:
Wode's Grimoire: Add particles to Flamestrike

PLEASE TEST THIS IN BUILD
The api need to include emitters for this to work
This also changes the code to something similar to bleed.
Flamestrike will now do the full emitter lifetime before letting others spells continue

BloodSpartan committed:
Wode's Grimoire: Add trails to Vengance.

Also provides a buff too, as caster will take damage first if targeted, then damage others
instead of dealing the same damage from start of cast
This also prevents targets taking different damage with same spell chain depending on who is targeted first

jordan-octavia-blue commented on Wodes_Grimoire/cards/Ensnare.ts:
```diff
@@ -25,57 +23,33 @@ const spell: Spell = {
-                    setTimeout(resolve, 100); //Very unfamiliar with promises and async functions. 
+                    setTimeout(resolve, 100);
                 })
             }
```
This whole block is unnecessary.  Promises and async are just ways of making something happen "later" rather than one line after another during code execution.

This block essentially says, "wait 100 milliseconds then return the function".  This was only necessary for Undead Blade and Slash to match the animation.

```
                            setTimeout(() => {
                                Unit.takeDamage(unit, damageDone, state.casterUnit, underworld, prediction, state);
                            }, 100)
```

In retrospect, Undead Blade is a little more complicated than most spells and I shouldn't have made it the first mod example.

TL;DR you can remove this block, it just makes the function wait 100 ms before returning.

jordan-octavia-blue commented on Wodes_Grimoire/cards/FlameStrike.ts:
Well done! As I see you did here, the prediction resolved immediately while the non-prediction resolves when the particles are done.  Fantastic!

jordan-octavia-blue commented on Wodes_Grimoire/cards/Pacify.ts:
Maybe Pacify could set `mana` and `maxMana` of casters to 0 so they can't cast too? Or is it only supposed to effect melee units?

BloodSpartan replied:
I have it to specifically not target the support class of units, like priest and summoner because their actions on their turn aren't technically "attacks", it should still work against ranged units like archer and glop. Setting mana/max to 0 as well would also mess with mana steal and burns.

jordan-octavia-blue replied:
Ah yes, that makes sense!

BloodSpartan committed:
Wode's Grimoire: Clean up Dominate code

Removing outdated comments and unneeded timouts

jordan-octavia-blue committed:
Wode's Grimoire: fix: Dominate

Removes extra loop.  Only play SFX once regardless of how many
targets there are.
Fix faction check

jordan-octavia-blue committed:
Wode's Grimoire: Clean Fast_Forward

Remove unneeded animationDelaySum and timeout

jordan-octavia-blue committed:
Wode's Grimoire: ref FlameStrike

Extract explosionTargets since it's the same in both places.
Replace quantity loop for quantity-modified damage variables. (The loop
was executing immediately and didn't result in any visual
difference anyway).

jordan-octavia-blue committed:
Wode's Grimoire: Add to description of Pacify

Auto Lint Harvest and Pacify

jordan-octavia-blue committed:
Wode's Grimoire: Extract common logic in Vengance

I pulled out the damage loop.  I find it best to extract shared logic so
that if it is ever changed or refactored it only has to change in one
place and nothing will be left behind unchanged.

Changed quantity loop to just multiplication

jordan-octavia-blue committed:
Wode's Grimoire: Fix FlameStrike

getUnitsWithinDistanceOfTarget will also return the target
if the target is a unit (target can be any Vec2).
So I had to refactor the damage logic.  Previously it
was dealing too much to the `unit` (40+10) and if
you used target circle + flamestrike and the circle
included the player it wouldn't deal damage to the player
at all (I think because of the .splice)

jordan-octavia-blue committed:
Wode's Grimoire: Add floating text to Grace and Regen

jordan-octavia-blue committed:
Wode's Grimoire: Improve description for Ensnare and Pacify

Auto lint

jordan-octavia-blue committed:
Wode's Grimoire: Add floating text to fast forward

jordan-octavia-blue committed:
Wode's Grimoire: Remove large wip icon

so that it doesn't get shipped

jordan-octavia-blue committed:
Wode's Grimoire: Resolve FlameStrikeWithParticles

so it doesn't hang on prediction or multiplayer

jordan-octavia-blue committed:
Wode's Grimoire: Fix Flamestrike hanging on server

It needs to resolve on server and prediction,
it was just resolving on prediction

jordan-octavia-blue committed:
core: Update mods to include spell icons

in spritesheets so the icons can be used in multiplayer thought bubbles
